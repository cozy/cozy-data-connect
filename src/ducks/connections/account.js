import {
  CREATE_CONNECTION,
  ENQUEUE_CONNECTION,
  PURGE_QUEUE,
  UPDATE_CONNECTION_ERROR,
  UPDATE_CONNECTION_RUNNING_STATUS
} from './'

const DOCTYPE = 'io.cozy.accounts'

const reducer = (state = {}, action) => {
  switch (action.type) {
    case CREATE_CONNECTION:
      return state
    case ENQUEUE_CONNECTION:
      return { ...state, isEnqueued: true }
    case PURGE_QUEUE:
      return { ...state, isEnqueued: false }
    case UPDATE_CONNECTION_ERROR:
      const { error } = action
      if (error) return { ...state, error }
      return state
    case UPDATE_CONNECTION_RUNNING_STATUS:
      const { isRunning } = action
      const hasRun = state.hasRun || (!!state.isRunning && !isRunning)
      return {
        ...state,
        isRunning,
        hasRun,
        error: isRunning ? null : state.error || null
      }
    default:
      return state
  }
}

export default reducer

// selectors
export const getAccount = (state, accountId) =>
  state.documents &&
  state.documents[DOCTYPE] &&
  state.documents[DOCTYPE][accountId]

export const getConnectionStatus = state => {
  if (hasError(state)) return 'error'
  if (isRunning(state)) return 'ongoing'
  if (hasRun(state)) return 'done'
  return 'pending'
}

const hasError = state => {
  return !isRunning(state) && !!state.error
}

const hasRun = state => {
  return !!state.hasRun
}

export const isEnqueued = state => {
  return !!state.isEnqueued
}

const isRunning = state => {
  return !!state.isRunning
}
