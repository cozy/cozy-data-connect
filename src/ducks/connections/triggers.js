export const reducer = (state = {}, action) => {
  switch (action.type) {
    case 'ENQUEUE_CONNECTION':
      return {
        ...state,
        [action.trigger._id]: {
          ...state[action.trigger._id],
          isEnqueued: true
        }
      }
    case 'LAUNCH_TRIGGER':
      return {
        ...state,
        [action.trigger._id]: {
          ...state[action.trigger._id],
          isRunning: true
        }
      }
    default:
      return state
  }
}

export default reducer
