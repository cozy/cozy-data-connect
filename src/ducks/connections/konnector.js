import {
  CREATE_CONNECTION,
  DELETE_CONNECTION,
  ENQUEUE_CONNECTION,
  PURGE_QUEUE,
  UPDATE_CONNECTION_ERROR,
  UPDATE_CONNECTION_RUNNING_STATUS
} from './'

import account, {
  getConnectionStatus,
  isQueued
} from './account'

import { getKonnectorIcon } from '../../lib/icons'

const reducer = (state = {}, action) => {
  switch (action.type) {
    case CREATE_CONNECTION:
    case ENQUEUE_CONNECTION:
    case UPDATE_CONNECTION_ERROR:
    case UPDATE_CONNECTION_RUNNING_STATUS:
      return { ...state, [action.account._id]: account(state[action.account._id], action) }
    case PURGE_QUEUE:
      return Object.keys(state).reduce((accounts, accountId) => {
        return { ...accounts, [accountId]: account(state[accountId], action) }
      }, state)
    case DELETE_CONNECTION:
      // mind = blown : https://stackoverflow.com/questions/36553129/what-is-the-shortest-way-to-modify-immutable-objects-using-spread-and-destructur
      return (({[action.account._id]: deleted, ...state}) => state)(state)
    default:
      return state
  }
}

export default reducer

export const getQueuedConnections = (state, registryKonnector) => {
  return Object.keys(state).reduce((runningConnections, accountId) => {
    const label = registryKonnector.name
    const status = getConnectionStatus(state[accountId])
    const icon = getKonnectorIcon(registryKonnector)
    return isQueued(state[accountId])
      ? runningConnections.concat({ label, status, icon })
        : runningConnections
  }, [])
}
