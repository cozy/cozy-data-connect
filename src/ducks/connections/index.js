import konnector, { getQueuedConnections } from './konnector'

// constant
export const CREATE_CONNECTION = 'CREATE_CONNECTION'
export const DELETE_CONNECTION = 'DELETE_CONNECTION'
export const ENQUEUE_CONNECTION = 'ENQUEUE_CONNECTION'
export const PURGE_QUEUE = 'PURGE_QUEUE'
export const UPDATE_CONNECTION_RUNNING_STATUS =
  'UPDATE_CONNECTION_RUNNING_STATUS'
export const UPDATE_CONNECTION_ERROR = 'UPDATE_CONNECTION_ERROR'

// reducers

const reducer = (state = {}, action) => {
  switch (action.type) {
    case CREATE_CONNECTION:
    case DELETE_CONNECTION:
    case ENQUEUE_CONNECTION:
    case UPDATE_CONNECTION_ERROR:
    case UPDATE_CONNECTION_RUNNING_STATUS:
      if (!action.konnector || !action.konnector.slug)
        throw new Error('Missing konnector slug')
      return {
        ...state,
        [action.konnector.slug]: konnector(state[action.konnector.slug], action)
      }
    case PURGE_QUEUE:
      return Object.keys(state).reduce((konnectors, slug) => {
        return { ...konnectors, [slug]: konnector(state[slug], action) }
      }, state)
    default:
      return state
  }
}

export default reducer

// action creators sync
export const createConnection = (konnector, account, folder) => ({
  type: CREATE_CONNECTION,
  konnector,
  account,
  folder
})

export const deleteConnection = (konnector, account) => ({
  type: DELETE_CONNECTION,
  konnector,
  account
})

export const enqueueConnection = (konnector, account) => ({
  type: ENQUEUE_CONNECTION,
  konnector,
  account
})

export const purgeQueue = () => ({
  type: PURGE_QUEUE
})

export const updateConnectionError = (konnector, account, error) => ({
  type: UPDATE_CONNECTION_ERROR,
  konnector,
  account,
  error
})

export const updateConnectionRunningStatus = (
  konnector,
  account,
  isRunning = false
) => ({
  type: UPDATE_CONNECTION_RUNNING_STATUS,
  konnector,
  account,
  isRunning
})

// selectors
export const getQueue = (state, konnectorsRegistry) => {
  // Todo: not sur if storing into data is a good idea.
  if (!state.data) return []
  return Object.keys(state.data).reduce((runningConnections, konnectorSlug) => {
    const konnectorAccounts = state.data[konnectorSlug]
    return runningConnections.concat(
      getQueuedConnections(konnectorAccounts, konnectorsRegistry[konnectorSlug])
    )
  }, [])
}
