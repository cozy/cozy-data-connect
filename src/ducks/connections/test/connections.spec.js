/* eslint-env jest */
import get from 'lodash/get'

import connections, {
  enqueueConnection,
  getConnectionsByKonnector,
  getQueue,
  purgeQueue,
  RECEIVE_DATA
} from '../'

describe('Connections Duck', () => {
  describe('Reducer', () => {
    describe('Receiving data', () => {
      it('should ignore actions without data', () => {
        const initialState = { creation: null, konnectors: {} }

        expect(
          connections(initialState, {
            type: RECEIVE_DATA
          })
        ).toEqual(initialState)
        expect(
          connections(initialState, {
            type: RECEIVE_DATA,
            response: {
              data: { yo: 1 }
            }
          })
        ).toEqual(initialState)
        expect(
          connections(initialState, {
            type: RECEIVE_DATA,
            response: {
              data: []
            }
          })
        ).toEqual(initialState)
      })

      it('should add a new trigger', () => {
        const initialState = { creation: null, konnectors: {} }
        const data = [
          {
            _id: 'trigger-id',
            _type: 'io.cozy.triggers',
            message: {
              konnector: 'my-kon',
              account: 'account 1'
            },
            current_state: {
              status: 'done',
              last_error: null,
              last_execution: '2019-01-01'
            }
          }
        ]
        const newState = connections(initialState, {
          type: RECEIVE_DATA,
          response: {
            data
          }
        })

        expect(Object.keys(newState.konnectors).length).toEqual(1)
        const triggerState = get(
          newState,
          'konnectors.my-kon.triggers.trigger-id'
        )
        expect(triggerState).toBeDefined()
        expect(triggerState.account).toBe('account 1')
        expect(triggerState.hasError).toBe(false)
        expect(triggerState.isRunning).toBe(false)
        expect(triggerState.lastSyncDate).toBe('2019-01-01')
      })

      it('should add a new job', () => {
        const initialState = { creation: null, konnectors: {} }
        const data = [
          {
            _id: 'job-id',
            _type: 'io.cozy.jobs',
            worker: 'konnector',
            trigger_id: 'job-trigger-id',
            state: 'done',
            error: null,
            queued_at: '2019-01-01',
            message: {
              konnector: 'my-kon'
            }
          }
        ]
        const newState = connections(initialState, {
          type: RECEIVE_DATA,
          response: {
            data
          }
        })

        expect(Object.keys(newState.konnectors).length).toEqual(1)
        const triggerState = get(
          newState,
          'konnectors.my-kon.triggers.job-trigger-id'
        )
        expect(triggerState).toBeDefined()
        expect(triggerState.account).toBe(undefined)
        expect(triggerState.hasError).toBe(false)
        expect(triggerState.isRunning).toBe(false)
        expect(triggerState.lastSyncDate).toBe('2019-01-01')
      })
    })

    describe('full trigger informations', () => {
      it('should add a new trigger', () => {
        const initialState = { creation: null, konnectors: {} }
        const data = [
          {
            _id: 'trigger-id',
            _type: 'io.cozy.triggers',
            message: {
              konnector: 'my-kon',
              account: 'account 1'
            },
            current_state: {
              status: 'done',
              last_error: null,
              last_execution: '2019-01-01'
            }
          }
        ]
        const newState = connections(initialState, {
          type: RECEIVE_DATA,
          response: {
            data
          }
        })

        expect(get(newState, 'konnectors.my-kon.triggers.data')).toEqual(data)
      })

      it('should keep the latest version of a trigger', () => {
        const initialState = {
          creation: null,
          konnectors: {
            'my-kon': {
              triggers: {
                data: [
                  {
                    _id: 'trigger-id',
                    _type: 'io.cozy.triggers',
                    message: {
                      konnector: 'my-kon',
                      account: 'account 1'
                    },
                    current_state: {
                      status: 'done',
                      last_error: null,
                      last_execution: '2019-01-01'
                    }
                  }
                ]
              }
            }
          }
        }

        const newConnectorData = {
          _id: 'trigger-id',
          _type: 'io.cozy.triggers',
          message: {
            konnector: 'my-kon',
            account: 'account 1'
          },
          current_state: {
            status: 'done',
            last_error: null,
            last_execution: '2019-02-01'
          }
        }

        const newState = connections(initialState, {
          type: RECEIVE_DATA,
          response: {
            data: [newConnectorData]
          }
        })

        const triggerState = get(newState, 'konnectors.my-kon.triggers.data')
        expect(triggerState.length).toEqual(1)
        expect(triggerState[0]).toEqual(newConnectorData)
      })

      it('should not add jobs to the list', () => {
        const initialState = {
          creation: null,
          konnectors: {
            'my-kon': {
              triggers: {
                data: [
                  {
                    _id: 'trigger-id',
                    _type: 'io.cozy.triggers',
                    message: {
                      konnector: 'my-kon',
                      account: 'account 1'
                    },
                    current_state: {
                      status: 'done',
                      last_error: null,
                      last_execution: '2019-01-01'
                    }
                  }
                ]
              }
            }
          }
        }

        const newConnectorData = {
          _id: 'job-id',
          _type: 'io.cozy.jobs',
          worker: 'konnector',
          trigger_id: 'trigger-id',
          state: 'done',
          error: null,
          queued_at: '2019-02-01',
          message: {
            konnector: 'my-kon'
          }
        }

        const newState = connections(initialState, {
          type: RECEIVE_DATA,
          response: {
            data: [newConnectorData]
          }
        })

        const triggerState = get(newState, 'konnectors.my-kon.triggers.data')
        expect(triggerState.length).toEqual(1)
        expect(triggerState[0].current_state).toBeDefined()
      })
    })
  })

  describe('Action creators', () => {
    describe('enqueueConnection', () => {
      it.skip('marks account as queued', () => {
        const state = {
          testprovider: {
            '17375ac5a59e4d6585fc7d1e1c75ec74': {}
          }
        }
        const konnector = { slug: 'testprovider' }
        const account = { _id: '17375ac5a59e4d6585fc7d1e1c75ec74' }

        const result = connections(state, enqueueConnection(konnector, account))

        expect(result).toMatchSnapshot()
      })
    })

    describe('purgeQueue', () => {
      it.skip('marks all accounts as not queued', () => {
        const state = {
          testprovider: {
            '17375ac5a59e4d6585fc7d1e1c75ec74': {},
            '63c670ea9d7b11e7b5888c88b1c12d46': {
              isEnqueued: true
            }
          },
          anotherprovider: {
            '768ccdaa9d7b11e7869aae88b1c12d46': {
              isEnqueued: true
            }
          }
        }

        const result = connections(state, purgeQueue())

        expect(result).toMatchSnapshot()
      })
    })
  })

  describe('Selectors', () => {
    describe('getConnectionsByKonnector', () => {
      it('returns expected connections', () => {
        const state = {
          konnectors: {
            provider: {
              triggers: {
                '81a548fca81455ec2c2644dd55009990': {
                  account: '81a548fca81455ec2c2644dd55008b52',
                  error: 'LOGIN_FAILED',
                  hasError: true,
                  isConnected: false,
                  isRunning: false
                },
                '63c670ea9d7b11e7b5888c88b1c12d46': {
                  account: '17375ac5a59e4d6585fc7d1e1c75ec74',
                  error: null,
                  hasError: false,
                  isConnected: true,
                  isRunning: true
                }
              }
            }
          }
        }

        const validKonnectors = ['provider']
        const validAccounts = ['81a548fca81455ec2c2644dd55008b52']

        expect(
          getConnectionsByKonnector(
            state,
            'provider',
            validAccounts,
            validKonnectors
          )
        ).toMatchSnapshot()
      })
    })

    describe('getQueue', () => {
      it('returns one queued connection per queued account', () => {
        const state = {
          data: {
            testprovider: {
              '17375ac5a59e4d6585fc7d1e1c75ec74': {},
              '63c670ea9d7b11e7b5888c88b1c12d46': {
                isRunning: true,
                isEnqueued: true
              },
              '768ccdaa9d7b11e7869aae88b1c12d46': {
                isEnqueued: true,
                error: {
                  message: 'test error'
                }
              }
            }
          }
        }

        const konnectors = {
          testprovider: {
            name: 'Test Provider',
            slug: 'testprovider'
          }
        }

        const result = getQueue(state, konnectors)

        expect(result).toMatchSnapshot()
      })
    })
  })
})
