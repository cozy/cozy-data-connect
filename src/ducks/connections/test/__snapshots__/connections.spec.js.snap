exports[`Connections Duck Action creators createConnection adds new connection 1`] = `
Object {
  "cozy": Object {
    "9bf93550308311c59f0a0047fc00fa1b": Object {},
  },
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {},
  },
}
`;

exports[`Connections Duck Action creators createConnection adds new connection in empty state 1`] = `
Object {
  "cozy": Object {
    "9bf93550308311c59f0a0047fc00fa1b": Object {},
  },
}
`;

exports[`Connections Duck Action creators deleteConnection deletes existing connection 1`] = `
Object {
  "testprovider": Object {
    "63c670ea9d7b11e7b5888c88b1c12d46": Object {},
  },
}
`;

exports[`Connections Duck Action creators enqueueConnection marks account as queued 1`] = `
Object {
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {
      "isEnqueued": true,
    },
  },
}
`;

exports[`Connections Duck Action creators purgeQueue marks all accounts as not queued 1`] = `
Object {
  "anotherprovider": Object {
    "768ccdaa9d7b11e7869aae88b1c12d46": Object {
      "isEnqueued": false,
    },
  },
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {
      "isEnqueued": false,
    },
    "63c670ea9d7b11e7b5888c88b1c12d46": Object {
      "isEnqueued": false,
    },
  },
}
`;

exports[`Connections Duck Action creators updateConnectionError set an error 1`] = `
Object {
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {
      "error": [Error: test error],
    },
  },
}
`;

exports[`Connections Duck Selectors getConnectionsByKonnector returns expected connections 1`] = `
Array [
  Object {
    "account": "81a548fca81455ec2c2644dd55008b52",
    "error": "LOGIN_FAILED",
    "hasError": true,
    "isConnected": false,
    "isRunning": false,
  },
]
`;

exports[`Connections Duck Selectors getQueue returns one queued connection per queued account 1`] = `Array []`;
